src/Common/GuardedPoolAllocator.cpp

```diff
@@ -0,0 +1,328 @@
+
+    // printf("\n");
+    return reinterpret_cast<void *>(user_ptr);
+}
+
+void GuardedPoolAllocator::trapOnAddress(uintptr_t address, Error err)
+{
+    state.failure_type = err;
+    state.failure_address = address;
+
+    /// Raise SIGSEGV by touching first guard page.
+    volatile char * p = reinterpret_cast<char *>(state.guarded_page_pool);
+    *p = 0;
+
+    __builtin_trap();
+}
+
+void GuardedPoolAllocator::deallocate(void * ptr)
+{
+    fflush(stdout);
+    uintptr_t u_ptr = reinterpret_cast<uintptr_t>(ptr);
+    size_t slot = state.getNearestSlot(u_ptr);
+    uintptr_t slot_start = state.slotToAddr(slot);
+    AllocationMetadata * meta = addrToMetadata(u_ptr);
+
+    if (meta->addr != u_ptr)
+    {
+        /// If multiple errors occur at the same time, use the first one.
+        std::lock_guard guard{pool_reservation_mutex};
+        trapOnAddress(u_ptr, Error::INVALID_FREE);
+    }
+
+    {
+        std::lock_guard guard{pool_reservation_mutex};
+        if (meta->is_deallocated)
+        {
+            trapOnAddress(u_ptr, Error::DOUBLE_FREE);
+        }
+
+        /// Ensure that the deallocation is recorded before marking the page as
+        /// inaccessible. Otherwise, a racy use-after-free will have inconsistent
+        /// metadata.
+        meta->recordDeallocation();
+
+        /// Ensure that the unwinder is not called if the recursive flag is set,
+        /// otherwise non-reentrant unwinders may deadlock.
+        if (!getThreadLocals()->recursive_guard)
+        {
+            ScopedRecursiveGuard recursive_guard;
+            meta->deallocation_trace.recordBacktrace();
+        }
+    }
+
+    deallocateInGuardedPool(reinterpret_cast<void *>(slot_start), state.maximumAllocationSize());
+
+    /// Acquire mutex to free slot
+    std::lock_guard guard{pool_reservation_mutex};
+    freeSlotInGuardedPool(slot);
+
+    /// I want to remove this output at the last moment before merge
+    // printf("Deallocated ptr=%p, slot=%zu", ptr, slot);
+    // printf("\n");
+}
+
+size_t GuardedPoolAllocator::reserveSlotInGuardedPool()
+{
+    /// Do not reuse slots until we've given out all initial slots
+    if (currently_allocated_slots < state.max_simultaneous_allocations)
+        return currently_allocated_slots++;
+
+    if (free_slots_length == 0)
+        return kInvalidSlotID;
+
+    size_t reserved_index = getRandomUint32T() % free_slots_length;
+    size_t slot_index = free_slots[reserved_index];
+    free_slots[reserved_index] = free_slots[--free_slots_length];
+    return slot_index;
+}
+
+void GuardedPoolAllocator::freeSlotInGuardedPool(size_t slot_index)
+{
+    assert(free_slots_length < state.max_simultaneous_allocations);
+    free_slots[free_slots_length++] = slot_index;
+}
+
+size_t GuardedPoolAllocator::getPlatformPageSize()
+{
+    /// Use function from ClickHouse base
+    return static_cast<size_t>(getPageSize());
+}
+
+uint32_t GuardedPoolAllocator::getRandomUint32T() const
```

Comment:
```suggestion
uint32_t GuardedPoolAllocator::getRandomUInt32T() const
```

Or simply `getRandom` ?